### PR TITLE
Ensure diagnostic pipeline is drained before Dispose() returns

### PR DESCRIPTION
**Pull request by karolz-ms**
This solves issue https://github.com/Azure/diagnostics-eventflow/issues/16

Also included in this change:
1. An update to Nuget package creation script (using PowerShell will make it easier to maintain going forward)

NOTE: this will require changing the TFS build definitions

2. Make sure the TraceInput is unregistered from the tracing infrastructure when disposing

### CODE REVIEW

File: src/Microsoft.Diagnostics.EventFlow.Core/Implementations/DiagnosticPipeline.cs [hunk 1]
```diff
@@ -156,11 +184,14 @@ public void Dispose()
 
             DisposeOf(this.inputSubscriptions);
 
-            this.pipelineHead.Complete();
-            this.pipelineHead.Completion.Wait(TimeSpan.FromMilliseconds(this.pipelineConfiguration.PipelineCompletionTimeoutMsec));
+            TimeSpan pipelineDrainWaitTime = PollWaitForPipelineDrain();
+
+            pipelineHead.Complete();
+            TimeSpan completionWaitTime = TimeSpan.FromMilliseconds(Math.Max(100, this.pipelineConfiguration.PipelineCompletionTimeoutMsec - pipelineDrainWaitTime.TotalMilliseconds));
```
Comment (brahmnes) [inline src/Microsoft.Diagnostics.EventFlow.Core/Implementations/DiagnosticPipeline.cs]: Should add a comment about why the 100 ms being the minimum waiting time.

File: src/Microsoft.Diagnostics.EventFlow.Core/Implementations/DiagnosticPipeline.cs [hunk 2]
```diff
@@ -229,16 +290,19 @@ public EventData[] FilterAndDecorate(EventData[] eventsToFilter)
                     return null;
                 }
 
-                // CONSIDER: using a Queue<T> with external lock and be able to reuse the queue
-                // (ConcurrentQueue has no Clear() method while ordinary Queue does).
-                ConcurrentQueue<EventData> eventsToKeep = new ConcurrentQueue<EventData>();
+                ConcurrentStack<EventData> eventsToKeep;
```
Comment (brahmnes) [inline src/Microsoft.Diagnostics.EventFlow.Core/Implementations/DiagnosticPipeline.cs]: Why switch from queue to stack? With the switch, the events are now returned in reverse order in the array. Does it matter?

File: src/Microsoft.Diagnostics.EventFlow.Core/Implementations/DiagnosticPipeline.cs [hunk 3]
```diff
@@ -188,24 +219,52 @@ private void DisposeOf(IEnumerable<object> items)
...
+        }
+
         private class FilterAction
         {
             private IReadOnlyCollection<IFilter> filters;
             private ParallelOptions parallelOptions;
             private ExecutionDataflowBlockOptions executionDataflowBlockOptions;
             private IHealthReporter healthReporter;
+            private Action<int> eventsRemovedNotification;
+            private ConcurrentBag<ConcurrentStack<EventData>> stackPool;
```
Comment (yantang-msft) [inline src/Microsoft.Diagnostics.EventFlow.Core/Implementations/DiagnosticPipeline.cs]: I think this deserves a comment

File: src/Microsoft.Diagnostics.EventFlow.Core/Implementations/DiagnosticPipeline.cs [hunk 4]
```diff
@@ -13,18 +13,20 @@
...
     {
-        private List<IDisposable> pipelineDisposables;
+        private List<IDisposable> pipelineLinkDisposables;
         private List<IDisposable> inputSubscriptions;
         private bool disposed;
         private bool disposeDependencies;
         private CancellationTokenSource cancellationTokenSource;
         private IDataflowBlock pipelineHead;
         private DiagnosticPipelineConfiguration pipelineConfiguration;
+        private volatile int eventsInProgress;
```
Comment (yantang-msft) [inline src/Microsoft.Diagnostics.EventFlow.Core/Implementations/DiagnosticPipeline.cs]: I feel it's hard to maintain this variable across blocks. Ideally it's supported by DataFlow natively.
Is it possible to have such a variable for each block, and dispose from the head to the tail?

File: src/Microsoft.Diagnostics.EventFlow.Core/Implementations/DiagnosticPipeline.cs [hunk 5]
```diff
@@ -97,14 +110,24 @@ public class DiagnosticPipeline : IDisposable
 
             if (sinks.Count > 1)
             {
+                // After broadcasting we will effectively have (sinks.Count - 1) more events in the pipeline.
+                var eventCounter = new TransformBlock<EventData[], EventData[]>(
+                    (batch) => { Interlocked.Add(ref this.eventsInProgress, (sinks.Count - 1) * batch.Length);  return batch; },
```
Comment (yantang-msft) [inline src/Microsoft.Diagnostics.EventFlow.Core/Implementations/DiagnosticPipeline.cs]: Can you add more comment on this?
This is what I see, and it's even.
NumObservedAtInput - NumDroppedAtFilter - NumOutput = 0

How does this eventCounter work in the calculation of eventInProgress?

File: src/Microsoft.Diagnostics.EventFlow.Core/Implementations/DiagnosticPipeline.cs (hunk 1 above)
Comment (karolz-ms) [inline src/Microsoft.Diagnostics.EventFlow.Core/Implementations/DiagnosticPipeline.cs]: I will add a comment. 

It is not just about draining the pipeline from events, but also about disposing of all internal resources that the pipeline blocks may use. Even if the pipeline is not fully drained (e.g. due to stuck filter/output), we still want to call Complete and give it some time.

File: src/Microsoft.Diagnostics.EventFlow.Core/Implementations/DiagnosticPipeline.cs (hunk 2 above)
Comment (karolz-ms) [inline src/Microsoft.Diagnostics.EventFlow.Core/Implementations/DiagnosticPipeline.cs]: We do not guarantee the order of the events in the pipeline will be maintained, so from correctness perspective the change is fine. Actually, there is no guarantee that events will be submitted to the pipeline in any particular order. Nor do we guarantee that the batches of events are sent out in order by outputs.

Each event has a timestamp so we know what happened when.

File: src/Microsoft.Diagnostics.EventFlow.Inputs.Trace/TraceInput.cs [hunk 6]
```diff
@@ -161,15 +162,35 @@ public IDisposable Subscribe(IObserver<EventData> observer)
...
         {
+            if (!this.disposed)
+            {
+                lock (this.subject)
+                {
+                    if (!this.disposed)
+                    {
+                        try
+                        {
+                            this.disposed = true;
```
Comment (karolz-ms) [inline src/Microsoft.Diagnostics.EventFlow.Inputs.Trace/TraceInput.cs]: Oops. Fixed

File: src/Microsoft.Diagnostics.EventFlow.Inputs.Trace/TraceInput.cs [hunk 7]
```diff
@@ -161,15 +162,35 @@ public IDisposable Subscribe(IObserver<EventData> observer)
...
+                        }
+                        finally
+                        {
+                            this.disposed = true;
+                        }
+                    }
+                }
+            }
+
             base.Dispose(disposing);
```
Comment (karolz-ms) [inline src/Microsoft.Diagnostics.EventFlow.Inputs.Trace/TraceInput.cs]: According to .NET design guidelines one should be able to call Dispose() multiple times on the same object, with no ill effects. In this particular piece of code we want to ensure that we call Trace.Listeners.Remove(this) _exactly once_ as part of disposing, regardless how many times Dispose() is called

File: src/Microsoft.Diagnostics.EventFlow.Core/Implementations/DiagnosticPipeline.cs (hunk 4 above)
Comment (karolz-ms) [inline src/Microsoft.Diagnostics.EventFlow.Core/Implementations/DiagnosticPipeline.cs]: Ideally, yes.

If we do not use completion propagation, complete from head to tail, and wait for completion on each block, that would give us what we want, i.e. drain the pipeline ensuring that each event is processed fully.

But that would be a bit tricky code to walk the pipeline and also account for how much time is left before we hit the timeout.

I think counting the events in the pipeline is simpler.

File: src/Microsoft.Diagnostics.EventFlow.Core/Implementations/DiagnosticPipeline.cs [hunk 8]
```diff
@@ -40,16 +42,26 @@ public class DiagnosticPipeline : IDisposable
...
             Requires.Argument(sinks.Count > 0, nameof(sinks), "There must be at least one sink");
 
+            this.eventsInProgress = 0;
+
             this.pipelineConfiguration = pipelineConfiguration ?? new DiagnosticPipelineConfiguration();
 
+            // An estimatie how many batches of events to allow inside the pipeline.
+            // We want to be able to process full buffer of events, but also have enough batches in play in case of high concurrency.
+            int MaxNumberOfBatchesInProgress = Math.Max(
+                5 * this.pipelineConfiguration.MaxConcurrency,
```
Comment (karolz-ms) [inline src/Microsoft.Diagnostics.EventFlow.Core/Implementations/DiagnosticPipeline.cs]: Yes, just a guess for reasonable size of the input buffer for the blocks that handler event batches.

File: src/Microsoft.Diagnostics.EventFlow.Core/Implementations/DiagnosticPipeline.cs [hunk 9]
```diff
@@ -116,25 +139,30 @@ public class DiagnosticPipeline : IDisposable
...
                     sink.Output,
                     this.cancellationTokenSource.Token,
-                    this.pipelineConfiguration.MaxEventBatchSize,
+                    MaxNumberOfBatchesInProgress,
                     this.pipelineConfiguration.MaxConcurrency,
-                    healthReporter);
-                ActionBlock<EventData[]> outputBlock = outputAction.GetOutputBlock();
-                this.pipelineDisposables.Add(outputSource.LinkTo(outputBlock, propagateCompletion));
+                    healthReporter,
+                    (eventsSentCount) => Interlocked.Add(ref this.eventsInProgress, -eventsSentCount));
```
Comment (karolz-ms) [inline src/Microsoft.Diagnostics.EventFlow.Core/Implementations/DiagnosticPipeline.cs]: Interlocked method do not use any locks.

I run the benchmark and there is no discernible effect on performance.

File: src/Microsoft.Diagnostics.EventFlow.Core/Implementations/DiagnosticPipeline.cs [hunk 10]
```diff
@@ -249,7 +313,15 @@ public EventData[] FilterAndDecorate(EventData[] eventsToFilter)
...
                 });
 
-                return eventsToKeep.ToArray();
+                EventData[] eventsFiltered = eventsToKeep.ToArray();
+                eventsToKeep.Clear();
+                this.stackPool.Add(eventsToKeep);
+                int eventsRemoved = eventsToFilter.Length - eventsFiltered.Length;
+                if (eventsRemoved > 0)
+                {
+                    this.eventsRemovedNotification(eventsRemoved);
```
Comment (karolz-ms) [inline src/Microsoft.Diagnostics.EventFlow.Core/Implementations/DiagnosticPipeline.cs]: That is ensured in the constructor

File: src/Microsoft.Diagnostics.EventFlow.Core/Implementations/DiagnosticPipeline.cs (hunk 5 above)
Comment (karolz-ms) [inline src/Microsoft.Diagnostics.EventFlow.Core/Implementations/DiagnosticPipeline.cs]: The broadcaster clones the events, that is why we need to increase the counter. Once you realize that it is actually pretty simple. I have updated the comment to make it clearer.

**Pull request merged**